designs/RestoreAndBuildCompatibility.md

```diff
@@ -0,0 +1,224 @@
+
+In this design, we are trying to improve the general mismatched tooling experience, but also consider alternatives for the particular issues customers are hitting with VS 16.8/NuGet 5.8/ SDK 5.0.100.
+
+## Who are the customers
+
+PackageReference customers
+
+## Goals
+
+* Improve the experience for mismatched versions of restore/build tooling.
+
+## Non-Goals
+
+* Introducing explicit backwards/forward compatibility. That's a lateral effort.
+
+## Solution Overview
+
+In the 2 particular problems described above the experience when things go wrong is inconsistent. It's not *always* apparent why something goes wrong.
+This proposal covers improving that warning/error experience.
+
+### NuGet output versioning
+
+When PackageReference restore is run, NuGet writes out the 3 files mentioned above. 2 of those files have a `version`.
+
+* The `project.assets.json` has a version that has historically been used for `reading`, not tooling compatibility checks.
+* The `nuget.g.props` contains a `NuGetToolVersion` property.
+
+### Build Tasks warning (SDK)
+
+Given that the implied compatibility exists, the proposal is for the SDK build tasks raise a suppressible warning when the output is not in the expected shape.
+
+* Warn when the restore output is from an older tool than anticipated.
+* Warn when the restore output is from a newer tool than anticipated.
+
+There are 2 technical options and each might have a different effect on the frequency in which this warning is raised.
+
+#### Option 1 - Use assets file version
+
+NuGet increments the `project.assets.json` version every time there's a functional change in behavior. The build tasks in the SDK are built with a specific version of NuGet, and the SDK could raise a warning when the `project.assets.json` version is not the one it expects.
+
+Cons:
+
+* This version was not used with tooling/restore meaning compatibility in mind, but rather for writing/reading purposes. While unlikely, it's possible that other tooling depends on it.
+
+Pros:
+
+* We would only increment this when a functional restore change is made.
+
+Alternatively we could add a new version property, but that's probably unnecessary.
+
+#### Option 2 - Use NuGet tool version
+
+The nuget.g.props write out a `NuGetToolVersion` property which contains the version of the NuGet tooling that generated it.
+
+Cons:
+
+* This has the potential to be too noisy. Just because the version change that doesn't mean it's not compatible.
+
+Pros:
+
+* The flip side is that we prefer that customers use consistent versions of the tooling, so going this direction has the added benefit of consolidating tooling for our customers.
+
+#### Option 3 - Define a new compatibility version
+
+Pros:
+
+* We would only increment this when a functional restore change is made.
+
+Cons:
+
+* Both NuGet and SDK tooling require changes to add the initial warning.

```

Comment:
Like the existing assets file version number proposal, this makes any assets file change difficult to insert.